turbopack/crates/turbo-tasks-macros/src/func.rs

```diff
@@ -109,61 +108,56 @@ impl TurboFn<'_> {
+                                         self: Vc<Self> instead",
+                                        definition_context.function_type(),
+                                    ))
+                                    .emit();
+                                return None;
+                            }
 
-                    match &reference {
-                        None => {
-                            receiver
-                                .span()
-                                .unwrap()
-                                .error(format!(
-                                    "{} cannot take self by value, use &self or self: Vc<Self> \
-                                     instead",
-                                    definition_context.function_type(),
-                                ))
-                                .emit();
-                            return None;
+                            match type_reference.elem.as_ref() {
+                                syn::Type::Path(TypePath { qself: None, path })
+                                    if path.is_ident("Self") => {}
+                                _ => {
+                                    self_type
+                                        .span()
+                                        .unwrap()
+                                        .error(
+                                            "Unexpected `self` type, use `&self` or `self: \
+                                             Vc<Self>",
+                                        )
+                                        .emit();
+                                    return None;
+                                }
+                            }
                         }
-                        Some((_, Some(lifetime))) => {
-                            lifetime
+                        syn::Type::Path(_) => {}
+                        _ => {
+                            self_type
                                 .span()
                                 .unwrap()
-                                .error(format!(
-                                    "{} cannot take self by reference with a custom lifetime, use \
-                                     &self or self: Vc<Self> instead",
-                                    definition_context.function_type(),
-                                ))
+                                .error("Unexpected `self` type, use `&self` or `self: Vc<Self>")
                                 .emit();
                             return None;
                         }
-                        _ => {}
                     }
+                    // We don't validate that the user provided a valid
+                    // `turbo_tasks::Vc<Self>` here.
+                    // We'll rely on the compiler to emit an error
+                    // if the user provided an invalid receiver type
+
+                    let ident: Ident = self_token.clone().into();
```

Comment:
Very tiny nit: Probably not everyone on the team feels the same, but I usually prefer using the `From` trait when the type isn't otherwise obvious. E.g. I tend to use `RcStr::from(...)`.

```suggestion
                    let ident = Ident::from(self_token.clone());
```